x/mongo/driver/topology/pool_test.go

```diff
@@ -445,11 +445,7 @@ func TestPool(t *testing.T) {
 		})
 	})
 	t.Run("checkOut", func(t *testing.T) {
-		t.Parallel()
```

Comment:
See this [patch](https://spruce.mongodb.com/task/mongo_go_driver_race_test__version~latest_os_ssl_40~ubuntu1804_64_go_1_18_test_race_patch_1c35f30cdc586dd9ea120cb3442fcada61abe5ba_637259e20ae606456f885cd8_22_11_14_15_08_26/tests?execution=0&sortBy=STATUS&sortDir=ASC). `TestPool` seemed to have data races with my other changes (particularly from `getOrQueueForIdleConn`); I investigated for about an hour but am still unsure why these data races are occurring. My changes to `TestPool` create a 1 second increase in overall testing time, so I'm not sure it's worth it to continue investigating unless one of you has an idea what might be happening.

Reply:
That race detector failures highlights an actual race condition bug in the `pool` test case `checkOut/closes_perished_connections` (see [here](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/pool_test.go#L516)). I suspect that the `assert.NotEqualf` call [here](https://github.com/mongodb/mongo-go-driver/blob/5edaf2b788defdb6fe888136dfc98b4acc074808/x/mongo/driver/topology/connection.go#L477) is actually using reflection to access the fields on the `*connection`, not just assert that the pointers are equal. That's causing unsynchronized access to the `*connection.state` field.

The fix for that race condition is to only assert that the pointers are not equal.
```go
// Assert that the connection pointers are not equal. Don't use "assert.NotEqual" because it asserts
// non-equality of fields, possibly accessing some fields non-atomically and causing a race condition.
assert.True(t, c1 != c2, "expected a new connection on 2nd check out after idle timeout expires")
```

Reply:
Ah great catch; that seems believable to me. I see no other instances of direct `assert` comparison between `*connection` in `TestPool` either. Added the parallelism back in and used the assertion you wrote above.